RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,332 @@
+6. Propagating subclass instances correctly also with operators, using
+   views/slices/indexing/etc.
+7. Preserve subclass attributes when using methods or views/slices/indexing.
+8. A way to insert code that operates on both functions and methods uniformly
+   (so we can write a single function that overrides all operators).
+
+Achieving interoperability with NumPy and adopting its array protocols is out
+of scope for this proposal and we propose to defer it to a later proposal.
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the
+   `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's
+   `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have
+a much more streamlined experience. Namely, the following code example will
+work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors the ability to also modify the
+results of methods and operators in `__torch_function__`, along with regular
+function calls, and to modify the result to their specific use-case, perform
+logging, or otherwise change the result or the action of the method. For
+example:
+
+```python
+import logging
+
+class LoggingTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        logging.info(f"func: {func.__name__}, args: {args!r}, kwargs: {kwargs!r}")
+        return super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, LoggingTensor)
+            ),
+            args,
+            kwargs
+        )
+```
+
+Assuming minimum logging level is set to `logging.INFO`, the following
+indicates the code run, with the logging output in the comments.
+
+```python
+t = LoggingTensor([1])
+
+t.sum()  # sum, (LoggingTensor([1]),), {}
+t[0]  # __getitem__, (LoggingTensor([1]), 0,), {}
+
+# This is already possible
+torch.sum(t)  # sum, (LoggingTensor([1]),), {}
+```
+
+To make the protocol operate only on functions rather than methods, one can
+check for `func not in type(self).__dict__.values()`. To check for operators
+and/or indexing, one can check `func.__name__.endswith("__")`.
+
+### Performance
+There are a few requirements for the performance of this proposal, when
+implemented:
+
+* No sub-100 ns deterioration per function call on `torch.Tensor` objects.
```

Comment:
This is phrased wrong. "No sub-100 ns" says more than 100 ns is okay. Propose this change:

1. Requirement is: "no deterioration for performance of `torch.Tensor` instances."
2. Under the requirements list say: "The first requirement cannot be fully met. Due to the need to add `self` as an extra parameter in argument parsing, we expect a small (sub 100 ns) impact for method calls (and no impact for function calls)".